File: scripts/query_markdown_index.mjs
#!/usr/bin/env node
/**
 * query_markdown_index.mjs — Search, scan, and focus the memory-decay index.
 *
 * Now matches against section headings and includes substring boost.
 */
import { existsSync, readFileSync } from 'fs';
import { join, resolve } from 'path';

const INDEX_DIRNAME = '.memory-decay';
const INDEX_FILE = 'index.json';
const TIER_WEIGHT = { fresh: 1.0, recent: 0.85, faded: 0.6, ghost: 0.3, expired: 0 };

function fail(msg) { console.error(msg); process.exit(1); }

function findMemoryRoot(start) {
  let cur = resolve(start || process.cwd());
  while (cur !== '/' && cur !== resolve(cur, '..')) {
    if (existsSync(join(cur, 'memory'))) return join(cur, 'memory');
    cur = resolve(cur, '..');
  }
  return join(resolve(start || process.cwd()), 'memory');
}

function loadIndex(memoryRoot) {
  const p = join(resolve(memoryRoot, '..'), INDEX_DIRNAME, INDEX_FILE);
  if (!existsSync(p)) fail(`index not found: ${p}\nRun: node scripts/sync_markdown_index.mjs ${memoryRoot}`);
  return JSON.parse(readFileSync(p, 'utf8'));
}

function tokenize(text) {
  if (!text) return [];
  const tokens = [];
  const eng = text.toLowerCase().match(/[a-z0-9_\-.]+/g);
  if (eng) tokens.push(...eng);
  const cjk = text.match(/[\u4e00-\u9fff]+/g);
  if (cjk) for (const seg of cjk) {
    for (let i = 0; i < seg.length - 1; i++) tokens.push(seg.slice(i, i + 2));
    for (const ch of seg) tokens.push(ch);
  }
  return tokens;
}

function termFreq(tokens) {
  const tf = {};
  for (const t of tokens) tf[t] = (tf[t] || 0) + 1;
  return tf;
}

function cosineSim(a, b) {
  let dot = 0, magA = 0, magB = 0;
  for (const t in a) { magA += a[t] ** 2; if (b[t]) dot += a[t] * b[t]; }
  for (const t in b) magB += b[t] ** 2;
  return (!magA || !magB) ? 0 : dot / (Math.sqrt(magA) * Math.sqrt(magB));
}

function substringBoost(query, text) {
  return (query && text.toLowerCase().includes(query.toLowerCase())) ? 0.3 : 0;
}

function scoreEntry(qTokens, entry, rawQuery = '') {
  const summaryText = entry.summary || '';
  const sectionText = entry.section || '';
  const domainText = entry.domain || '';
  const summaryScore = cosineSim(qTokens, termFreq(tokenize(summaryText)));
  const sectionScore = cosineSim(qTokens, termFreq(tokenize(sectionText))) * 2;
  const domainScore = cosineSim(qTokens, termFreq(tokenize(domainText))) * 3;
  const subBoost = rawQuery
    ? Math.max(substringBoost(rawQuery, summaryText), substringBoost(rawQuery, sectionText))
    : 0;
  const tierW = TIER_WEIGHT[entry.tier] || 0.5;
  return (summaryScore + sectionScore + domainScore + subBoost) * tierW;
}

function search(index, query, limit = 8) {
  const q = termFreq(tokenize(query));
  const out = [];
  for (const e of index) {
    if (e.tier === 'expired') continue;
    const score = scoreEntry(q, e, query);
    if (score > 0) out.push({ ...e, score });
  }
  out.sort((a, b) => b.score - a.score);
  return out.slice(0, limit);
}

function scan(index, query) {
  const q = termFreq(tokenize(query));
  const domains = {};
  for (const e of index) {
    if (e.tier === 'expired') continue;
    const d = e.domain || 'uncategorized';
    if (!domains[d]) domains[d] = { domain: d, count: 0, latest: null, maxScore: 0 };
    domains[d].count++;
    const score = scoreEntry(q, e, query);
    if (score > domains[d].maxScore) domains[d].maxScore = score;
    if (!domains[d].latest || (e.created && e.created > domains[d].latest)) domains[d].latest = e.created;
  }
  return Object.values(domains).filter(d => d.maxScore > 0).sort((a, b) => b.maxScore - a.maxScore);
}

function focus(index, domain) {
  return index.filter(e => e.domain === domain && e.tier !== 'expired')
    .sort((a, b) => (b.created || '').localeCompare(a.created || ''));
}

const args = process.argv.slice(2);
if (!args.length) fail('Usage: node query_markdown_index.mjs search|scan|focus <arg>');
const [command, arg] = args;
const index = loadIndex(findMemoryRoot());

if (command === 'search') {
  if (!arg) fail('Query required.');
  for (const e of search(index, arg))
    console.log(`[${e.tier}] ${e.domain} | ${e.summary} | ${e.source} | ${(e.score*100).toFixed(0)}%`);
} else if (command === 'scan') {
  if (!arg) fail('Query required.');
  for (const d of scan(index, arg))
    console.log(`${d.domain} (${d.count} memories, latest: ${(d.latest||'?').slice(0,10)}, relevance: ${(d.maxScore*100).toFixed(0)}%)`);
} else if (command === 'focus') {
  if (!arg) fail('Domain required.');
  for (const e of focus(index, arg))
    console.log(`[${e.tier}] ${e.created.slice(0,10)} | ${e.summary} | ${e.source}`);
} else fail(`Unknown command: ${command}`);

File: scripts/query_markdown_index.py
#!/usr/bin/env python3
import json
import math
import os
import re
import sys
from pathlib import Path

INDEX_DIRNAME = '.memory-decay'
INDEX_FILE = 'index.json'
TIER_WEIGHT = {'fresh': 1.0, 'recent': 0.85, 'faded': 0.6, 'ghost': 0.3, 'expired': 0}


def fail(msg):
    print(msg, file=sys.stderr)
    sys.exit(1)


def find_memory_root(start=None):
    current = Path(start or os.getcwd()).resolve()
    for candidate in [current] + list(current.parents):
        if (candidate / 'memory').is_dir():
            return candidate / 'memory'
    return current / 'memory'


def index_path(memory_root):
    return memory_root.parent / INDEX_DIRNAME / INDEX_FILE


def load_index(path):
    if not path.exists():
        fail(f'index not found: {path}\nRun: python3 scripts/sync_markdown_index.py {path.parent.parent / "memory"}')
    return json.loads(path.read_text(encoding='utf8'))


def tokenize(text):
    if not text:
        return []
    tokens = []
    tokens.extend(re.findall(r'[a-z0-9_\-.]+', text.lower()))
    for seg in re.findall(r'[\u4e00-\u9fff]+', text):
        for i in range(len(seg) - 1):
            tokens.append(seg[i:i+2])
        tokens.extend(seg)
    return tokens


def term_freq(tokens):
    tf = {}
    for t in tokens:
        tf[t] = tf.get(t, 0) + 1
    return tf


def cosine_sim(a, b):
    dot = sum(a[t] * b.get(t, 0) for t in a)
    mag_a = sum(v * v for v in a.values())
    mag_b = sum(v * v for v in b.values())
    return 0 if not mag_a or not mag_b else dot / (math.sqrt(mag_a) * math.sqrt(mag_b))


def score_entry(q_tokens, entry):
    summary_score = cosine_sim(q_tokens, term_freq(tokenize(entry.get('summary', ''))))
    domain_score = cosine_sim(q_tokens, term_freq(tokenize(entry.get('domain', '')))) * 3
    tier_w = TIER_WEIGHT.get(entry.get('tier'), 0.5)
    return (summary_score + domain_score) * tier_w


def search(index, query, limit=8):
    q = term_freq(tokenize(query))
    out = []
    for entry in index:
        if entry.get('tier') == 'expired':
            continue
        score = score_entry(q, entry)
        if score > 0:
            item = dict(entry)
            item['score'] = score
            out.append(item)
    out.sort(key=lambda x: x['score'], reverse=True)
    return out[:limit]


def focus(index, domain):
    out = [e for e in index if e.get('domain') == domain and e.get('tier') != 'expired']
    out.sort(key=lambda x: x.get('created', ''), reverse=True)
    return out


def scan(index, query):
    q = term_freq(tokenize(query))
    domains = {}
    for entry in index:
        if entry.get('tier') == 'expired':
            continue
        domain = entry.get('domain', 'uncategorized')
        domains.setdefault(domain, {'domain': domain, 'count': 0, 'latest': None, 'maxScore': 0})
        domains[domain]['count'] += 1
        score = score_entry(q, entry)
        domains[domain]['maxScore'] = max(domains[domain]['maxScore'], score)
        created = entry.get('created')
        if not domains[domain]['latest'] or (created and created > domains[domain]['latest']):
            domains[domain]['latest'] = created
    return sorted([d for d in domains.values() if d['maxScore'] > 0], key=lambda x: x['maxScore'], reverse=True)


def main():
    if len(sys.argv) < 2:
        fail('Usage: python3 scripts/query_markdown_index.py search <query> | scan <query> | focus <domain>')

    command = sys.argv[1]
    arg = sys.argv[2] if len(sys.argv) > 2 else None
    memory_root = find_memory_root()
    index = load_index(index_path(memory_root))

    if command == 'search':
        if not arg:
            fail('Query required.')
        for e in search(index, arg):
            print(f"[{e['tier']}] {e['domain']} | {e['summary']} | {e['source']} | {e['score'] * 100:.0f}%")
    elif command == 'scan':
        if not arg:
            fail('Query required.')
        for d in scan(index, arg):
            latest = (d['latest'] or '?')[:10]
            print(f"{d['domain']} ({d['count']} memories, latest: {latest}, relevance: {d['maxScore'] * 100:.0f}%)")
    elif command == 'focus':
        if not arg:
            fail('Domain required.')
        for e in focus(index, arg):
            print(f"[{e['tier']}] {e['created'][:10]} | {e['summary']} | {e['source']}")
    else:
        fail(f'Unknown command: {command}')


if __name__ == '__main__':
    main()
